deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AutoPromoter.java

```diff
@@ -252,66 +357,60 @@ void processOnce(String envId) throws Exception {
+            String schedule = promoteBean.getSchedule();
+            if (!StringUtils.isEmpty(schedule)) {
+                if (!autoDeployDue(currDeployBean, schedule)) {
+                    return;
+                }
+                // For cron auto promote, we only choose the latest one.
+                size = Math.min(size, 1);
+            }
 
-        // Otherwise, promote from pred env deploys
-        promoteDeploy(currEnvBean, currDeployBean, size, promoteBean);
+            // Otherwise, promote from pred env deploys
+            promoteDeploy(currEnvBean, currDeployBean, size, promoteBean);
+        }
     }
 
     boolean autoDeployDue(DeployBean deployBean, String cronExpressionString) {
         Date date = new Date();
+        return !date.before(getScheduledCheckDueTime(deployBean.getStart_date(), cronExpressionString));
+    }
+
+    public Date getScheduledCheckDueTime(long start_date, String cronExpressionString){
+        Date ret = new Date(0);
         try {
             if (!CronExpression.isValidExpression(cronExpressionString)) {
                 LOG.error(String.format("Cron expression %s is not valid. Ignore it.", cronExpressionString));
-                return true;
+                return ret;
             }
             CronExpression cronExpression = new CronExpression(cronExpressionString);
-            if (deployBean == null) {
-                return true;
-            }
-            Date lastDeloyDate = new Date(deployBean.getStart_date());
-            Date nextDeployDate = cronExpression.getNextValidTimeAfter(lastDeloyDate);
-            // Only run the cron deploy when the current date is equal or after the scheduled deploy date
-            // since last deploy.
-            //
-            // If current date is before the next scheduled deploy date since last deploy, return false.
-            if (date.before(nextDeployDate)) {
-                LOG.info(String.format("The next scheduled deploy after last deploy %tc is %tc, now is: %tc",
-                    nextDeployDate, lastDeloyDate, date));
-                return false;
-            } else {
-                return true;
-            }
+            Date lastCheckDate = new Date(start_date);
+            ret = cronExpression.getNextValidTimeAfter(lastCheckDate);
+            return ret;
         } catch (ParseException e) {
             LOG.error(String.format("Failed to parse cron expression: %s. Reason: %s",
                 cronExpressionString, e.getMessage()));
-            return true;
+            return ret;
         } catch (Exception e) {
             LOG.error(String.format("Failed to validate date. Reason: %s", e.getMessage()));
-            return true;
+            return ret;
         }
     }
 
-    BuildBean getBuildCandidate(EnvironBean envBean, long after, int size) throws Exception {
+    BuildBean getBuildCandidate(EnvironBean envBean, long before, long after, int size) throws Exception {
```

Comment:
before and after is confusing. Based on the logic of the method, it really means an interval [after,  before]. Further, people's brain thinks of after first, then before.